### DISCUSSION TITLE
I appreciate your work!

### DISCUSSION THREAD

**Discussion by GwnDaan** [Open-Agriculture/AgIsoStack-plus-plus#12] [General]
Hey!

This project is definitely promising and you've already done a great job! I can't wait for a stable version to come out. In the meantime I'd be willing to help where I can, like some testing or documentation? Although I'm not really proficient in C++ I don't mind learning a few things along the way. I have a few terminals at home with isobus that I can do some testing on if that helps. Looking forward to see the open source community for isobus grow!

**Comment by ad3154** [comment]
Hey there! Thanks! I _really_ appreciate your support! I would also like to see the open source community for ISOBUS and related technologies (like RTK, implement rate control, etc) grow! I think that there are a lot of people out there that want to get involved with the technologies but struggle with getting going because most software and hardware that would help them are proprietary, expensive, or highly technical. I'm hoping that this project will one day be a reliable, stable starting point for those folks so they don't need to focus on as much "the basics" to get a project started.

If you've got some terminals and are able to do some testing from time-to-time, that might be super helpful actually. My resources are pretty limited. It's on my to-do list to develop some test applications that upload an object pool to a terminal as examples for how to use the library, which would be amazing to test against real displays. In theory, you could test it from any linux PC with a CAN adapter, even a raspberry pi would work. The linux limitation is due to my need for socket CAN today, but if you are limited to Windows or something I could probably add support for some specific CAN adapter driver as well. So, if you're open to that, I will work on some of those example programs!

**Reply by GwnDaan** [reply]
Dope! I'm definitely willing to test out whatever it might be! How I see all the technology around me is exactly as you describe, and I think with open source a lot can be less expensive and more usable, because the basics will already be there and resources can be used for innovation instead. 

On a hobby level, with a raspberry pi and a canbus module I could test things. I have 2 different terminal types, one built in by Fendt and the other is a standalone Müller terminal. I think neither one is above VT version 4, but I'm not sure. They both have section control and a task controller. But those things would be for later. I'll happily invest my time in testing and debugging this project. Looking forward to see the example programs you can come up with! With some documentation on the implemented object pool functionality/specification I can also try to build a few complete testing pools to use. Just sharing some idea's here haha, lmk your thoughts.

**Reply by GwnDaan** [reply]
Actually, I have 3 terminals with the oldest being >10 or even >14 years old lol. Also built by Müller and probably something like VT version 1 or 2. I just completely forgot it with the better technology out there today, we still use it for a sprayer that doesn't work with the newer VT versions somehow.

**Reply by ad3154** [reply]
I haven't forgotten about our discussion regarding testing, by the way! If you're still open to it, I am working on a VT example program that goes more in depth than just the object pool upload, it has just been on the back burner for a while. It turns out, creating IOP files is a bit tough when you're just an open source guy and not a company, but I'm working through it! I think a comprehensive example is important for long term success. Anyways, thank you again for all your support @GwnDaan

**Reply by GwnDaan** [reply]
@ad3154 Dope! I had been busy lately but I have managed to test out the example and was surprised by how easy it was to run straight out of the box! Good job there! I came across a few issues which I raised in the GitHub's issue tracker. Other than that, everything from alarm mask to the buttons, it all just worked flawlessly. 

I plan to develop a Virtual Terminal application this winter to use during next planting season for controlling an implement. The automatic actions would be based on the tractor's NMEA2000 messages and the provided configuration in the VT. Do you recon this library is ready for an implementation like this, or do you recommend me to look elsewhere? It is for personal use and not necessary for the implement to work, so it does not need to be as reliable. I feel like it might be a good test for the library. Let me know what you think.

**Reply by ad3154** [reply]
That is cool! I think the VT portion of the library is pretty close and could probably be used with minimal issues, but I have not started NEMA2K fast packet support yet. I don't think it would be too tough to implement, so if that is a priority for you I can put it at the top of the list.

**Reply by ad3154** [reply]
I've gone ahead and added NMEA 2000 support in PR #69 but I don't have a ton of ways to test it, so any feedback you might have would certainly be welcome. @GwnDaan

**Reply by GwnDaan** [reply]
Thanks! I'll start to work on the project and raise issues for any issues I run into. Feel free to close any that don't bring enough value to the project. I'll also check if I can get some testing done for the N2K FP implementation. Though I don't think I'll need to the FP during the project as I only need the XTE message, which is afaik a regular can message.

**Reply by GwnDaan** [reply]
> I've gone ahead and added NMEA 2000 support in PR #69 but I don't have a ton of ways to test it, so any feedback you might have would certainly be welcome. @GwnDaan

The receiving part works good as far as I can tell! The data seems to get through successfully: https://github.com/ad3154/ISO11783-CAN-Stack/commit/921d652458b285623fc5b89f3dc444400af6fcb9#commitcomment-92942699. But I wonder if [this code](https://github.com/GwnDaan/ISO11783-CAN-Stack/blob/9492d52ff71f3bbfc94ee728e6b669ecd4a6f816/examples/nmea2000/main.cpp#L25-L30) can be written any more human friendly. It seems like it's prone to errors when reading larger fields, e.g. the altitude of 8 bytes.
I'm not sure how to test the sending part though.

**Comment by Thom-de-Jong** [comment]
This project really is awesome!

I have also been developing an ISOBUS stack in my spare time and might be able to help out here.
My cpp is a bit rusty as I have been using Rust for my stack. I realy think combining all of our forces into one project would greatly speed up open-source isobus development.
I will set up a cpp dev environment and try it out!

**Reply by ad3154** [reply]
Hey there! I've always been a bit interested in learning rust, that's great! I just mostly live in C++ at my job, so it was the most comfortable for me. 

I am a bit surprised with how many people out there actually care about ISOBUS, haha! It's awesome to see.

I would certainly welcome contributions to this project. In fact it would probably help discover issues, keep me motivated, and highlight any improvements we need in documentation to get people going. For example, a contributor probably needs to know how the network manager and the generic "protocol" class work at a high level. So those could really use some good, human generated documentation probably.

Anyways, please be sure to let me know if you have any issues when trying out this stack, or if anything is super unclear. My goal here is to make things as easy and reliable as possible. Checking out the example programs might help while I finish up the official tutorial webpage (Coming soon™).

**Comment by rahibd** [comment]
Hi, I am a student at Dalhousie University. I am going to install and test this program soon on an RPi4 + CAN HUT. I am trying to mane my own ECU which will take some parameters from the VT (Raven CR7 as VT.) for actuation. The actuation part is done so far but just started the VT part. I think this will be a great test case for your project as well. 
Cheers.

**Comment by emorthen** [comment]
Hey there!
Is it possible to say when this software will be mature enough to serve as the basis for commercial ISOBUS applications? I see commits with bug fixes, however "Stable Release" still seems to be on the roadmap. Asking from the space of real time weed detection and precision spraying 🌿

Regardless of the above; thank you for an amazing job and effort. And if I am not mistaken, the ISOBUS community will benefit greatly from such a democratization of this knowledge and experience.

Rock on, guys🚜💪

**Reply by GwnDaan** [reply]
Hey @emorthen welcome! I'd say it is mature enough to some extend, there is still quite possible that you encounter bugs, but we're actively developing the project so it won't take long to get them fixed after it's reported. On the other hand, since it is completely open-source, you also have a nice opportunity to have a shot at fixing the bug yourself and not rely others :)

**Reply by emorthen** [reply]
Thank you for your response, appreciate it 🙏 We are considering to take the step to dig into the substance of ISOBUS, hopefully we will then contribute back to this community!

**Comment by 0x-a6** [comment]
Awesome project, I'm a cpp engineer working at a agritech startup - I'd like to get involved. I don't have a wide range of hardware, but I can get my hands on a lot of different equipment if needed. :wave:

**Comment by Berduino** [comment]
Great project, nice work! I do have medium experience with Python, but C++/cmake and compiling is new for me. Trying to figuring it out. The tutorials and stack overflow are very useful, but still I am walking into 'new' errors with each step. I will make it, I think, but if there is someone that knows about a video or so for a very comparable project to really get started with the basics of Cmake in combination with C++ and CMD, I am happy to hear about that.

Furthermore, I am working with RAUCH/KUHN/CCI and I am able to test all the new (and old) software/hardware/object pools/aux functions of basically all machines. I do have different ISOBUS set-ups from full TECU/NMEA1083 simulator or just a plain connection of USB - PEAK - Power - ECU and CCI 1200 terminal.

So really looking forward being able to work with this and being able to help you guys. For now, I am using my time to get things up and running. ;)

**Reply by GwnDaan** [reply]
Hey there, great to hear about your interest in the project! If there's anything you encounter that we can assist you with don't hesitate to ask for help 😄. 

> but still I am walking into 'new' errors with each step

It would be really valuable to us if you would be able to enumerate which errors you bumped into that were hard to find in the tutorial / stackoverflow. Most likely others with less prior experience will encounter it as well, we try to keep the things people tend to mess up abstract and simple, but from our side it's hard to guess which questions people could have. So any assistance you can give there will be really appreciated!

>if there is someone that knows about a video or so for a very comparable project to really get started with the basics of Cmake in combination with C++ and CMD, I am happy to hear about that.

This section of our contribution guide might be valuable to you: https://github.com/Open-Agriculture/AgIsoStack-plus-plus/blob/main/CONTRIBUTING.md#setting-up-a-development-environment-with-gui. It should set you up with a ready-to-go environment for testing our samples. I'm not sure about other projects with the basics explained but I expect there to be tons of guides online.

> I am able to test all the new (and old) software/hardware/object pools/aux functions of basically all machines. I do have different ISOBUS set-ups from full TECU/NMEA1083 simulator or just a plain connection of USB - PEAK - Power - ECU and CCI 1200 terminal.

That's great to hear! We have a few samples you can try in the [examples](https://github.com/Open-Agriculture/AgIsoStack-plus-plus/tree/main/examples) folder you might be able to try out. We are working on a few more examples as well, keep an eye out for those

Lastly, we have a discord and telegram group you can also join for getting tech support while you are setting things up, and for good vibes in general :)

**Reply by Berduino** [reply]
Hi GwnDaan, Thanks for your reply! I will let you know if I get stuck and stackoverflow/google seems not able to help me. ;) How can I add myself/be added to the Telegram chat? Would be interesting. Furthermore, for any questions regarding CCI 800/1200/A3/.., I am more then happy to help.

**Reply by Berduino** [reply]
found the telegram group :)
